### ISSUE TITLE
Shorthand syntax for conversions in overloads

### ISSUE THREAD

**Issue by jckarter** [jckarter/clay#102]
A common boilerplate pattern is a pair of overloads, one that converts values to a canonical type and another that actually applies an algorithm to the canonical type:

```
foo(x, y) = foo(Int(x), String(y));

overload foo(x:Int, y:String) { ... }
```

It would be nice to have this built into the language. An idea KS and I came up with was to reuse the `as` keyword. The following definition:

```
foo(x as Int, y as String) { ... }
```

would be equivalent to the following two definitions:

```
foo(x, y) = forward ..foo(x as Int, y as String);

overload foo(x:Int, y:String) { ... }
```

The new `as` operator should be distinct from the constructor function, since operations like `int as Int` should be no-ops, and the implicit conversion operation should be limited to cheap and safe value-to-argument conversions in keeping with the "explicit is better than implicit" philosophy.

**Comment by ghost** [comment]
How about:

``` text
  foo(x:TX as Int, y:TY as String) {
      /* body of foo */
  }
```

desugers as:

``` text

  [TX, TY | Convertible?(TX, Int) and Convertible?(TY, String)]
  inline foo(forward x: TX, forward y: TY) = ...forward foo(TX, TY, x as Int, y as String);

  foo(TX, TY, x: Int,  y: String) {
      /* body of foo */
  }
```

Thus allowing access to the original types in the body of foo. This would also look nicer when procedures allow static parameters. Would this work?

**Comment by jckarter** [comment]
@agemogolk That makes sense. A definition with `as` could then apply a predicate or pattern restriction to the input types prior to conversion too, such as `[T | Integer?(T)] foo(x:T as Int) { ... }`.

**Comment by ghost** [comment]
Implementing this as an overload de-sugar could be problematic with respect to overload resolution and final overloads. Maybe it would be better to integrate argument promotion into the, yet to be decided, more sophisticated resolution scheme. Provide an overload-able `Convertible?` primitive etc. Although this would probably end up being similar to VB overloads . . . not sure how well that would work. 
Also, I suppose i'm not entirely convinced that new syntax should be introduced for a pattern that could be handled with  predicates (or type-classes) e.g.

```
[T, S  when Convertible?(T, Int) and Convertible?(S, String)]
foo(x:T, y:S) {
    var i, s = Int(x), String(y);
    /* body of foo */
}
```

**Comment by jckarter** [comment]
The problem with the naive implementation you described is that it instantiates `/* body of foo */` for every set of input types `T` and `S`, even though the generated code will be exactly the same for every instance at that point. This is a common enough pattern that I think having convenient syntax to do it the right way is desirable.

**Comment by ghost** [comment]
Ah yes I forgot about multiple instantiations! Hmm.

**Comment by jckarter** [comment]
Recent versions of GCC are able to do partial inlining of template instantiations when they recognize that the set of live values in a template has become monomorphic; the Clay compiler could implement a similar optimization, but such an optimization is still brittle and easy to accidentally disable.

**Comment by ghost** [comment]
Implemented in #461.
